### PR TITLE
Introduce `core.gardener.cloud/v1.ControllerDeployment` API

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area dev-productivity usability
/kind enhancement

**What this PR does / why we need it**:

This PR introduces the new `core.gardener.cloud/v1` API version including the `ControllerDeployment` resource.
The v1 API drops support for custom types and replaces `providerConfig` with a proper API structure for helm-based `ControllerDeployments`.
gardener-apiserver stores the `ControllerDeployment` resource in the v1 version but can perform lossless conversion from and to `v1beta1`.

The new API version is introduced to support further developments augmenting helm-based `ControllerDeployments` (e.g., OCI registry support instead of raw charts).
As we move away from the `providerConfig` approach, we need a new API version.

All other `core.gardener.cloud` resources are still only available in `v1beta1`. This PR lays the foundation for graduating other resources to `v1` later on as well.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/9773

**Special notes for your reviewer**:

Co-Authored-By: @maboehm 

**Release note**:

```feature operator
A new `core.gardener.cloud/v1` API version is introduced which only includes the `ControllerDeployment` resource for now. The new version of the `ControllerDeployment` drops the `type` and `providerConfig` fields in favor of a well-structured section for helm-based `ControllerDeployments`.
```
```breaking developer
`ControllerDeployment` objects using a custom type (other than `helm`) are deprecated. Support for custom types will be removed when the `core.gardener.cloud/v1beta1` API version is dropped.
```
```noteworthy developer
The `hack/generate-controller-registration.sh` script now generates a `ControllerDeployment` object in the `core.gardener.cloud/v1` API version.
```

### CODE REVIEW

File: pkg/apis/core/types_controllerdeployment.go [hunk 1]
```diff
@@ -34,3 +45,13 @@ type ControllerDeploymentList struct {
 	// Items is the list of ControllerDeployments.
 	Items []ControllerDeployment
 }
+
+// HelmControllerDeployment configures how an extension controller is deployed using helm.
+type HelmControllerDeployment struct {
+	// RawChart is the base64-encoded, gzip'ed, tar'ed extension controller chart.
+	// +optional
```
Comment (rfranzke) [inline pkg/apis/core/types_controllerdeployment.go]: Remove `// +optional` tags from internal API

File: pkg/apis/core/types_controllerdeployment.go [hunk 2]
```diff
@@ -34,3 +45,13 @@ type ControllerDeploymentList struct {
...
+	RawChart []byte
+	// Values are the chart values.
+	// +optional
```
Comment (rfranzke) [inline pkg/apis/core/types_controllerdeployment.go]: Remove `// +optional` tags from internal API

**Review by rfranzke — COMMENTED** [review]
Love it :) Looking forward. Thank you very much!

**Comment by timebertt** [comment]
Thanks, done. PTAL :)

**Comment by timebertt** [comment]
> PR needs rebase.

Done :)

**Comment by rfranzke** [comment]
I think we are good to go. Let's get this in, especially since it's a prerequisite for other features (cc @maboehm @kon-angelo @vknabel). Thanks for your efforts 🎉 

/lgtm
/approve

**Pull request merged**